src/utilities/xml/XMLValidator.cpp

```diff
@@ -0,0 +1,332 @@
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#include "XMLValidator.hpp"
+#include "XMLErrors.hpp"
+#include "XMLUtils.hpp"
+
+#include <libxml/xmlversion.h>
+#include <libxml/xmlreader.h>
+#include <libxml/parser.h>
+#include <libxml/tree.h>
+#include <libxml/schematron.h>
+
+#include <libxslt/xslt.h>
+#include <libxslt/xsltInternals.h>
+#include <libxslt/transform.h>
+#include <libxslt/xsltutils.h>
+#include <libxml/xpath.h>
+#include <libxml/xpathInternals.h>  // BAD_CAST
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <iterator>
+
+namespace openstudio {
+
+XMLValidator::XMLValidator(const openstudio::path& schemaPath, const openstudio::path& xmlPath)
+  : m_schemaPath(openstudio::filesystem::system_complete(schemaPath)), m_xmlPath(openstudio::filesystem::system_complete(xmlPath)) {
+
+  m_logSink.setLogLevel(Warn);
+  m_logSink.setChannelRegex(boost::regex("openstudio\\.XMLValidator"));
+  m_logSink.setThreadId(std::this_thread::get_id());
+
+  if (!openstudio::filesystem::exists(schemaPath)) {
+    LOG_AND_THROW("Schema '" << toString(schemaPath) << "' does not exist");
+  } else if (!openstudio::filesystem::is_regular_file(schemaPath)) {
+    LOG_AND_THROW("Schema '" << toString(schemaPath) << "' cannot be opened");
+  }
+
+  if (!openstudio::filesystem::exists(xmlPath)) {
+    LOG_AND_THROW("XML File '" << toString(xmlPath) << "' does not exist");
+  } else if (!openstudio::filesystem::is_regular_file(xmlPath)) {
+    LOG_AND_THROW("XML File '" << toString(xmlPath) << "' cannot be opened");
+  }
+
+  if (schemaPath.extension() == ".xsd") {
+    m_validatorType = XMLValidatorType::XSD;
+    LOG(Info, "Treating schema as a regular XSD");
+  } else if (schemaPath.extension() == ".xslt") {
+    m_validatorType = XMLValidatorType::XSLTSchematron;
+    LOG(Info, "Treating schema as a XLST StyleSheet that derives from a Schematron");
+  } else if ((schemaPath.extension() == ".xml") || (schemaPath.extension() == ".sct")) {
+    m_validatorType = XMLValidatorType::Schematron;
+    LOG(Info, "Treating schema as a Schematron");
+    LOG_AND_THROW("Opening a Schematron document isn't supported as the moment. Use python's lxml to extract your Schematron to an XSLT Stylesheet. "
+                  "```\n"
+                  "from lxml.isoschematron import Schematron;\n"
+                  "s = Schematron(file='schematron.sct', store_xslt=True)\n"
+                  "with open('schematron.xslt', 'w') as f:\n"
+                  "    f.write(str(s._validator_xslt))\n"
+                  "```");
```

Comment:
As I had noted on https://github.com/jmarrec/TestCpp-xml-schematron

* `libxml2` includes an old version of schematron. So the idea is to convert the schematron xml to an XSLT stylesheet, and use `libxslt` to apply that stylesheet and get validation errors.
    * this is what the python `lxml` module ends up doing.

### TODO:

* Currently I manually converted the schematron xml to an xslt stylesheet using `lxml`.

Reply:
lxml converts the schematron (or embedded schematron) to an XSLT by making a series of XSLT transforms: https://github.com/lxml/lxml/blob/master/src/lxml/isoschematron/__init__.py

We could probably replicate that (with a huge amount of pain for limited usage...).